OTEP.md

```diff
@@ -32,7 +32,415 @@ Using a configuration model or configuration file, users could configure all opt
 * Versioning: needs to be able to version stability while evolving
 * (stretch) The file format can be validated client side.
 
-## Internal details (TBD)
+## Internal details
+
+Each language implementation supporting OpenTelemetry *MUST* produce a model that is available for applications to leverage. This allows implementations to provide a configuration interface without the expectation on users to load a configuration file. The following demonstrates how Python and Go may provide a configuration interface to accomplish this:
```

Comment:
Is this necessary? If someone wants to do programatic configuration, why not skip this step and directly build the SDK?

Reply:
It's a good question. I guess since we're providing a data model for configuration, it seems like it should be available for use from the code, but maybe not?

I imagine this to be syntactic sugar to reduce the surface area of what folks interact with when configuring OTel.

Reply:
If I were a user I'd minimally want the following options:

```
// Read config file to string
String config = Files.readString(Paths.get("/config.yml"));

// Interpret config and return configured SDK
OpenTelemetrySdk sdk = OpenTelemetryFileConfig.configure(config);

// ..or just pass the path to the file
OpenTelemetry sdk = OpenTelemetryFileConfig.configure(Paths.get("/config.yml"));
```

Maybe you'd also want to be able to validate a file is syntactically correct and not configure an SDK. Could have validate methods for that:
```
try {
  OpenTelemetryFileConfig.validate(config);
} catch (Exception e) {
  e.printStackTrace();
}
```

Having a programatic representation of the config schema like you've suggested would be a natural response from a validate call. Could then pass it to a configure method:
```
// Read config file to string
String config = Files.readString(Paths.get("/config.yml"));

// Validate the config and parse to the programatic representation
OpenTelemetryFileConfig parsedConfig;
try {
  parsedConfig = OpenTelemetryFileConfig.parseAndValidate(config);
} catch (Exception e) {
  e.printStackTrace();
}

// Configure SDK fro programatic representation
OpenTelemetrySdk = OpenTelemetryFileConfig.configure(parsedConfig);
```

Perhaps the programatic representation is useful in some languages, but probably not all. In java, it reads very similarly to configuring the SDK programmatically, and feels redundant. Maybe it can be a MAY instead of a MUST?

Reply:
I've updated the MUST to a MAY. Thinking about this some more though, since instrumentations may use this configuration as well, I'm leaning towards the SDK and the instrumentations providing a method to configure themselves via a configuration model.


Something like this:

```
try {
  cfg = OpenTelemetryFileConfig.parseAndValidate(filename)
} catch (Exception e) {
  e.printStackTrace();
  exit();
}

OpenTelemetrySDK sdk = sdk.WithConfiguration(cfg)
OpenTelemetryInstrumentationX x = x.WithConfiguration(cfg)
```

Having the configuration method return the SDK would cause any SDK to have to reimplement the configuration parsing. I guess maybe that's desirable as they may have different configuration requirements?

Though for instrumentations I wouldn't expect each instrumentation to have to re-implement configuration parsing. Similarly, I wouldn't expect the configuration parsing to know how to instantiate every type of instrumentation.

Reply:
Hadn't thought about the instrumentation side of things. Makes sense to have some sort of `cfg` object / class that can be passed around so it doesn't have to be reparsed. 👍

Reply:
Added more details to the PR. PTAL @jack-berg @srikanthccv